src/main/scala/org/graphframes/GraphFrame.scala

```diff
@@ -466,8 +465,15 @@ class GraphFrame private(
       indexedVertices.select(
         col(ATTR + "." + ID).cast("long").as(LONG_ID), col(ATTR + "." + ID).as(ID), col(ATTR))
     } else {
-      val indexedVertices = zipWithUniqueId(vertices)
-      indexedVertices.select(col("uniq_id").as(LONG_ID), col("row." + ID).as(ID), col("row").as(ATTR))
+      val withLongIds = vertices.select(ID)
+        .distinct()
```

Comment:
Does distinct() do sorting or hashing already?  Could you hash partition and then implement distinct() within each partition?  I'm not sure how smart Catalyst is here.